opnode/rollup/driver/step.go

```diff
@@ -7,53 +7,119 @@ import (
+	FetchReceipts(ctx context.Context, id eth.BlockID) ([]*types.Receipt, error)
+	// FetchBatches from the given window of L1 blocks
+	FetchBatches(ctx context.Context, window []eth.BlockID) ([]derive.BatchData, error)
+	// FetchL2Info fetches the L2 header information corresponding to a L2 block ID
+	FetchL2Info(ctx context.Context, id eth.BlockID) (derive.L2Info, error)
+}
+
 type DriverAPI interface {
 	l2.EngineAPI
 	l2.EthBackend
 }
 
-// step takes an L1 block id and then creates and executes the L2 block. The forkchoice is updated as well.
-func (d *Driver) step(ctx context.Context, l1Input eth.BlockID, l2Parent eth.BlockID, l2Finalized common.Hash) (eth.BlockID, error) {
-	logger := d.log.New("input_l1", l1Input, "input_l2_parent", l2Parent, "finalized_l2", l2Finalized)
+type outputImpl struct {
+	dl     Downloader
+	rpc    DriverAPI
+	log    log.Logger
+	Config rollup.Config
+}
+
+// DriverStep derives and processes one or more L2 blocks from the given sequencing window of L1 blocks.
+// An incomplete sequencing window will result in an incomplete L2 chain if so.
+//
+// After the step completes it returns the block ID of the last processed L2 block, even if an error occurs.
+func (d *outputImpl) step(ctx context.Context, l2Head eth.BlockID, l2Finalized eth.BlockID, l1Input []eth.BlockID) (out eth.BlockID, err error) {
+	if len(l1Input) == 0 {
+		return l2Head, fmt.Errorf("empty L1 sequencing window on L2 %s", l2Head)
+	}
+
+	logger := d.log.New("input_l1_first", l1Input[0], "input_l1_last", l1Input[len(l1Input)-1],
+		"input_l2_parent", l2Head, "finalized_l2", l2Finalized)
+	logger.Trace("Running update step on the L2 node")
+
+	epoch := rollup.Epoch(l1Input[0].Number)
 
 	fetchCtx, cancel := context.WithTimeout(ctx, time.Second*20)
 	defer cancel()
-	bl, receipts, err := d.dl.Fetch(fetchCtx, l1Input)
+
+	l2Info, err := d.dl.FetchL2Info(fetchCtx, l2Head)
+	if err != nil {
+		return l2Head, fmt.Errorf("failed to fetch L2 block info of %s: %v", l2Head, err)
+	}
+	logger.Trace("Got l2 info")
+	l1Info, err := d.dl.FetchL1Info(fetchCtx, l1Input[0])
 	if err != nil {
-		return eth.BlockID{}, fmt.Errorf("failed to fetch block with receipts: %v", err)
+		return l2Head, fmt.Errorf("failed to fetch L1 block info of %s: %v", l1Input[0], err)
 	}
-	logger.Debug("fetched L1 data for driver")
+	logger.Trace("Got l1 info")
+	receipts, err := d.dl.FetchReceipts(fetchCtx, l1Input[0])
+	if err != nil {
+		return l2Head, fmt.Errorf("failed to fetch receipts of %s: %v", l1Input[0], err)
+	}
+	logger.Trace("Got receipts")
+	// TODO: with sharding the blobs may be identified in more detail than L1 block hashes
+	batches, err := d.dl.FetchBatches(fetchCtx, l1Input)
+	if err != nil {
+		return l2Head, fmt.Errorf("failed to fetch batches from %s: %v", l1Input, err)
+	}
+	logger.Trace("Got batches")
 
-	attrs, err := derive.PayloadAttributes(bl, receipts)
+	attrsList, err := derive.PayloadAttributes(&d.Config, l1Info, receipts, batches, l2Info)
```

Comment:
Batch parsing and filtering needs to happen somewhere between the downloader and the continued derivation of L2 blocks. We should probably fetch the list of transactions per L1 block instead, and then call the parsing/filtering here, to not leak that into the downloader logic.

When we introduce blob-transactions we'll end up needing an additional `FetchBlobs(ctx, versionedHashesList)` method. We still need to parse/filter transactions in that case, but instead of immediately parsing calldata we'll need to fetch the blobs referenced by the transaction, and then parse those into batches.

So now we change:
- `FetchBatches` -> `FetchTransactions`
- Call `BatchesFromEVMTransactions` after fetching
- Call `FilterBatches` after parsing

And later with blobs that will be:
- `FetchTransactions`
- `BatchBlobsFromEVMTransactions`
- `BatchesFromBlobs`
- `FilterBatches`

And depending on cost/performance trade-offs we might still encode e.g. the batch epoch and timestamp in calldata, but put all the other data in the blobs. That way we can skip out-of-range data easily

Reply:
Updated to follow specs. please double check.